server/lib/chat/chat-models.ts

```diff
@@ -578,3 +579,94 @@ export async function findChannelByName(
     done()
   }
 }
+
+/** Returns the number of channels that match a particular condition. */
+async function getChannelsCount(
+  whereCondition?: SQLStatement,
+  withClient?: DbClient,
+): Promise<number> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = sql`
+      SELECT COUNT(*) AS count
+      FROM channels
+    `
+
+    if (whereCondition) {
+      query.append(whereCondition)
+    }
+
+    const result = await client.query<{ count: string }>(query)
+
+    return Number(result.rows[0].count)
+  } finally {
+    done()
+  }
+}
+
+/**
+ * Gets a list of chat channels for a particular user, optionally filtered by a `searchStr`. If the
+ * user is an admin, this method will return a full list of channels. Otherwise, this method will
+ * only return non-private channels and channels that the user is not joined in.
+ */
+export async function getChannelsForUser(
+  userId: SbUserId,
+  isAdmin: boolean,
+  limit: number,
+  pageNumber: number,
+  searchStr?: string,
+  withClient?: DbClient,
+): Promise<{ channels: ChannelInfo[]; total: number }> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = isAdmin
+      ? sql`
+        SELECT c.*, COUNT(cu.user_id) AS user_count
+        FROM channels c
+        INNER JOIN channel_users cu ON c.id = cu.channel_id
+      `
+      : sql`
+        WITH unjoined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id NOT IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*, COUNT(cu.user_id) AS user_count
+        FROM channels c
+        INNER JOIN unjoined_channels ON c.id = unjoined_channels.channel_id
+        INNER JOIN channel_users cu ON c.id = cu.channel_id
+      `
+
+    let whereCondition: SQLStatement | undefined
+    if (!isAdmin) {
+      whereCondition = sql`WHERE private = false`
+    }
+    if (searchStr) {
+      const escapedStr = `%${searchStr.replace(/[_%\\]/g, '\\$&')}%`
+
+      if (whereCondition) {
+        whereCondition.append(sql` AND name ILIKE ${escapedStr}`)
+      } else {
+        whereCondition = sql`WHERE name ILIKE ${escapedStr}`
+      }
+
+      query.append(whereCondition)
+    }
+
+    query.append(sql`
+      GROUP BY c.id
+      ORDER BY user_count DESC, name
+      LIMIT ${limit}
+      OFFSET ${pageNumber * limit}
+    `)
+
+    const result = await client.query<DbChannel>(query)
+
+    const channels = await Promise.all(result.rows.map(row => convertChannelFromDb(row)))
+    const total = await getChannelsCount(whereCondition, withClient)
```

Comment:
This isn't a safe thing to do, the count can change between the original query you made and this query (and this also needlessly goes to the database twice for this information). Do you really need this information? It makes the query monumentally more expensive (the query now has to run against the entire table each time you grab a page, even though it may be able to deliver the page from a much smaller subset of the data) and it doesn't seem to be particularly useful in any UI

Reply:
Hmm, the only thing I need the total count for is for the infinite-loading-list in the UI (rendered through a carousel), to know when it reached the end. I'll try to think of a better way to implement pagination for that.

Reply:
You can do it like we do for chat messages:
1) Load a page
2) Does it have the amount of entries we asked for?
  - Yes: We still have more pages
  - No: We're done